### PR TITLE
header: fix authentication when protected header is zero-length map

### PR DESCRIPTION
**Pull request by pulsastrix**
COSE allows an empty protected header to be encoded as a zero-length map, even though the standard encourages encoding empty protected headers as a zero-length string (RFC 2119 SHOULD according to RFC 9052, Section 3).

However, according to RFC 9052, Section 4.4, 5.3 and 6.3, even if the header is encoded as a zero-length map, the structure used for authentication should not include the empty map if the protected header is empty ("if there are no protected attributes, a zero-length byte string is used").

Due to this, authentication of some of the official examples in the [cose-wg/Examples repository](https://github.com/cose-wg/Examples) was not possible using coset.
An example of this is the CoseSign1 object provided in
[sign1-tests/sign-pass-03.json](https://github.com/cose-wg/Examples/blob/master/sign1-tests/sign-pass-03.json), which uses the protected header encoding as a zero-length map.

This PR ensures that the correct behavior is implemented in coset by no longer including a zero-length map protected header during authentication.

Note: I am unsure as to how to proceed with the tests that now fail, as it seems like those explicitly expect a behavior different from the RFC.

### CODE REVIEW

File: src/header/mod.rs [hunk 1]
```diff
@@ -385,6 +385,19 @@ impl ProtectedHeader {
         ))
     }
 
+    pub fn to_be_authenticated(self) -> Result<Value> {
+        let result = self.cbor_bstr()?;
+        // protected header might have been encoded as a zero length map, only containing
+        // the byte 0xA0 (see RFC 9052, Section 3).
+        // However, this byte should not be included during authentication (RFC 9052, Section 4.4,
+        // 5.3 and 6.3), so we need to return an empty byte string here instead.
```
Comment (daviddrysdale) [inline src/header/mod.rs:393]: When would `ProtectedHeader::cbor_bstr()` ever return an (encoded) empty map?  

I _think_ the `is_empty()` check would mean that this could only happen if the original input (in `ProtectedHeader::original_data`) had the non-minimal encoding, and that object then got re-encoded using the original data.  Or is there another way it can happen?

Comment (pulsastrix) [reply src/header/mod.rs:393]: That should be the only way that this can happen, yes.

For actually re-encoding an existing COSE object this should not be an issue (hence why `cbor_bstr()` is unchanged).

However, as far as I understand the COSE specification for generating the structures that are provided to the signature/encryption/MAC-function (e.g. `Sig_structure`), this non-minimal encoding needs to be [explicitly replaced with a zero-length byte string](https://datatracker.ietf.org/doc/html/rfc9052#section-4.4-2.2) if no attributes are set, regardless of whether the actual encoding of the protected header is minimal or not.

The aforementioned examples/test cases explicitly check for this behavior (e.g. [sign1-tests/sign-pass-01](https://github.com/cose-wg/Examples/blob/master/sign1-tests/sign-pass-01.json) and [sign-tests/sign-pass-01](https://github.com/cose-wg/Examples/blob/master/sign-tests/sign-pass-01.json)), and in my own tests verification of these (supposedly valid) COSE structures only works with the changes made in this PR.

**Comment by daviddrysdale** [comment]
Coming back to this (apologies for the delay), I notice that there's an [open issue](https://github.com/cose-wg/Examples/issues/107)  that possibly indicates that the [sign1-tests/sign1-pass-01.json](https://github.com/cose-wg/Examples/blob/master/sign1-tests/sign-pass-01.json) test case is possibly expected to be invalid?

**Comment by pulsastrix** [comment]
> Coming back to this (apologies for the delay), I notice that there's an [open issue](https://github.com/cose-wg/Examples/issues/107) that possibly indicates that the [sign1-tests/sign1-pass-01.json](https://github.com/cose-wg/Examples/blob/master/sign1-tests/sign-pass-01.json) test case is possibly expected to be invalid?

Hm, the issue you mentioned does seem to refer to the same underlying issue, but I _think_ that the test case is actually correct.

There is at least one additional passage in the RFC that at least strongly implies that the protected bucket should not be used verbatim if it is a zero-length map:
> Senders SHOULD encode a zero-length map as a zero-length byte string rather than as a zero-length map (encoded as h'a0'). The zero-length byte string encoding is preferred, because it is both shorter **and the version used in the serialization structures for cryptographic computation**.

[RFC 9052, section 3](https://datatracker.ietf.org/doc/html/rfc9052#section-3-4.2.1)

This seems to imply that the zero-length map should not be used when creating `Sig_structure`/`Enc_structure`/`Mac_structure` instances, even if it is the encoding that was used when encoding the protected bucket.

Also, this being expected behavior would explain why the test case explicitly makes this change to the protected bucket, but isn't marked as an expected failure [^1].

Regardless of what the standard says, it might also be worth investigating how other COSE implementations handle this specific case.
I'll do some testing with other COSE implementations once I have the time to do so and report back.

[^1]: Some notes regarding the JSON test cases, because they are IMO very confusing and not well documented: 
         While [the JSON field that makes this header change](https://github.com/cose-wg/Examples/blob/53c9d634333bb4f529d78f5980fffa2667ee2c12/sign1-tests/sign-pass-01.json#L21) is itself part of the `failures` JSON object, the actual field that indicates an expected failure is a top level field called `fail` (see [this example in another test case](https://github.com/cose-wg/Examples/blob/53c9d634333bb4f529d78f5980fffa2667ee2c12/sign1-tests/sign-fail-07.json#L3)). 
         The other field being called `failures` seems to be a misnomer; Its contents are just more generally used to introduce modifications to the COSE structure after its creation, regardless of whether they should cause a failure or not.
         See also [test case sign-pass-03](https://github.com/cose-wg/Examples/blob/53c9d634333bb4f529d78f5980fffa2667ee2c12/sign1-tests/sign-pass-03.json#L23), which also has a "failure" that should be introduced, but isn't actually expected to fail.

**Comment by pulsastrix** [comment]
Hi, sorry for taking almost an entire year to get back to this.

I have written some minimal reproducible examples for implementations in `pycose`, `cose-java`, and `coset` to compare their behavior in this specific case, you can find the code for these implementations [here](https://github.com/pulsastrix/cose-empty-header-tests).

Using [the sign-pass-01 test case mentioned above](https://github.com/cose-wg/Examples/blob/53c9d634333bb4f529d78f5980fffa2667ee2c12/sign1-tests/sign-pass-01.json#L30C15-L30C211), which encodes the protected header as a zero length map, I get the following results:

Python (w. `pycose`):
```
➜ python verify_cose_message.py D28441A0A201260442313154546869732069732074686520636F6E74656E742E584087DB0D2E5571843B78AC33ECB2830DF7B6E0A4D5B7376DE336B23C591C90C425317E56127FBE04370097CE347087B233BF722B64072BEB4486BDA4031D27244F
<COSE_Sign1: [{}, {'Algorithm': 'Es256', 'KID': b'11'}, b'This ' ... (20 B), b'\x87\xdb\r.U' ... (64 B)]>
Signature valid: True
```

Java (w. `cose-java`):
```
➜ java -jar target/cose_empty_header_test_java-1.0-SNAPSHOT.jar D28441A0A201260442313154546869732069732074686520636F6E74656E742E584087DB0D2E5571843B78AC33ECB2830DF7B6E0A4D5B7376DE336B23C591C90C425317E56127FBE04370097CE347087B233BF722B64072BEB4486BDA4031D27244F
Validation successful: true
```

Rust (w. `coset`):
```
➜ cargo run -- D28441A0A201260442313154546869732069732074686520636F6E74656E742E584087DB0D2E5571843B78AC33ECB2830DF7B6E0A4D5B7376DE336B23C591C90C425317E56127FBE04370097CE347087B233BF722B64072BEB4486BDA4031D27244F
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/cose-empty-header-test-rust D28441A0A201260442313154546869732069732074686520636F6E74656E742E584087DB0D2E5571843B78AC33ECB2830DF7B6E0A4D5B7376DE336B23C591C90C425317E56127FBE04370097CE347087B233BF722B64072BEB4486BDA4031D27244F`
Verification result: Err(signature::Error {})
```

As a control, I have also tested all three implementations with a freshly generated COSE message that encodes the protected header as a zero length string (i.e., as recommended but not mandated by RFC 9052, Section 3):

Python (`pycose`):
```
➜ python verify_cose_message.py d28440a204423131012654546869732069732074686520636f6e74656e742e584087db0d2e5571843b78ac33ecb2830df7b6e0a4d5b7376de336b23c591c90c425317e56127fbe04370097ce347087b233bf722b64072beb4486bda4031d27244f
<COSE_Sign1: [{}, {'KID': b'11', 'Algorithm': 'Es256'}, b'This ' ... (20 B), b'\x87\xdb\r.U' ... (64 B)]>
Signature valid: True
```

Java (`cose-java`):
```
➜ java -jar target/cose_empty_header_test_java-1.0-SNAPSHOT.jar d28440a204423131012654546869732069732074686520636f6e74656e742e584087db0d2e5571843b78ac33ecb2830df7b6e0a4d5b7376de336b23c591c90c425317e56127fbe04370097ce347087b233bf722b64072beb4486bda4031d27244f
Picked up _JAVA_OPTIONS: -Djava.util.prefs.userRoot=/home/hugo/.config/java
Validation successful: true
```

Rust (`coset`):
```
➜ cargo run -- d28440a204423131012654546869732069732074686520636f6e74656e742e584087db0d2e5571843b78ac33ecb2830df7b6e0a4d5b7376de336b23c591c90c425317e56127fbe04370097ce347087b233bf722b64072beb4486bda4031d27244f
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/cose-empty-header-test-rust d28440a204423131012654546869732069732074686520636f6e74656e742e584087db0d2e5571843b78ac33ecb2830df7b6e0a4d5b7376de336b23c591c90c425317e56127fbe04370097ce347087b233bf722b64072beb4486bda4031d27244f`
Verification result: Ok(())
```

In summary, it seems that unlike `coset`, the other two implementations I tested (with one of them [being published by the COSE-WG](https://github.com/cose-wg/COSE-JAVA/)) indeed use a zero-length byte string for the protected header when constructing the `Sig_structure`, even if the protected header is not actually serialized as a zero-length string, but as a zero-length map.
